pixy/consumer.go

```diff
@@ -0,0 +1,562 @@
+		exclusiveConsumers:    make(map[string]map[int32]*exclusiveConsumer),
+		addTopicConsumerCh:    make(chan *topicConsumer),
+		deleteTopicConsumerCh: make(chan *topicConsumer),
+		stoppingCh:            make(chan none),
+	}
+	gc.dispatcher = newDispatcher(gc.baseCID, gc, sc.config)
+	return gc
+}
+
+func (gc *groupConsumer) String() string {
+	return gc.baseCID.String()
+}
+
+func (gc *groupConsumer) addTopicConsumer() chan<- *topicConsumer {
+	return gc.addTopicConsumerCh
+}
+
+func (gc *groupConsumer) deleteTopicConsumer() chan<- *topicConsumer {
+	return gc.deleteTopicConsumerCh
+}
+
+func (gc *groupConsumer) key() string {
+	return gc.group
+}
+
+func (gc *groupConsumer) start(stoppedCh chan<- dispatchTier) {
+	spawn(&gc.wg, func() {
+		defer func() { stoppedCh <- gc }()
+		gc.registry = spawnConsumerGroupRegister(gc.group, gc.config.ClientID, gc.config, gc.kazooConn)
+		var manageWg sync.WaitGroup
+		spawn(&manageWg, gc.managePartitions)
+		gc.dispatcher.start()
+		<-gc.stoppingCh
+		gc.dispatcher.stop()
+		gc.registry.stop()
+		manageWg.Wait()
+	})
+}
+
+func (gc *groupConsumer) requests() chan<- consumeRequest {
+	return gc.dispatcher.requests()
+}
+
+func (gc *groupConsumer) stop() {
+	close(gc.stoppingCh)
+	gc.wg.Wait()
+}
+
+func (gc *groupConsumer) dispatchKey(req consumeRequest) string {
+	return req.topic
+}
+
+func (gc *groupConsumer) newDispatchTier(key string) dispatchTier {
+	tc := gc.newTopicConsumer(key)
+	return tc
+}
+
+func (gc *groupConsumer) managePartitions() {
+	cid := gc.baseCID.NewChild("managePartitions")
+	defer cid.LogScope()()
+
+	topicConsumers := make(map[string]*topicConsumer)
+	var topics []string
+	var memberSubscriptions map[string][]string
+	ok := true
+	var nilOrRetryCh <-chan time.Time
+	var nilOrRegistryTopicsCh chan<- []string
+	shouldRebalance, canRebalance := false, true
+	rebalanceResultCh := make(chan error, 1)
+rebalancingLoop:
+	for {
+		select {
+		case tc := <-gc.addTopicConsumerCh:
+			topicConsumers[tc.topic] = tc
+			topics = listTopics(topicConsumers)
+			nilOrRegistryTopicsCh = gc.registry.topics()
+			continue rebalancingLoop
+		case tc := <-gc.deleteTopicConsumerCh:
+			delete(topicConsumers, tc.topic)
+			topics = listTopics(topicConsumers)
+			nilOrRegistryTopicsCh = gc.registry.topics()
+			continue rebalancingLoop
+		case nilOrRegistryTopicsCh <- topics:
+			nilOrRegistryTopicsCh = nil
+			continue rebalancingLoop
+		case memberSubscriptions, ok = <-gc.registry.membershipChanges():
+			if !ok {
+				goto done
+			}
+			nilOrRetryCh = nil
+			shouldRebalance = true
+		case err := <-rebalanceResultCh:
+			canRebalance = true
+			if err != nil {
+				log.Errorf("<%s> rebalance failed: err=(%s)", cid, err)
+				nilOrRetryCh = time.After(gc.config.Consumer.BackOffTimeout)
+				continue
+			}
+		case <-nilOrRetryCh:
+			shouldRebalance = true
+		}
+
+		if shouldRebalance && canRebalance {
+			topicConsumerCopy := make(map[string]*topicConsumer, len(topicConsumers))
```

Comment:
This wasn't obvious to me at first, so it might be worth a comment:

```go
// Copy topicConsumers to make sure rebalance doesn't see any changes we make while it is running.
```

That said, how expensive is `rebalance`? If there are a lot of topics, could this map copy take longer than `rebalance` itself?